python/snewpy/models/extended.py

```diff
@@ -0,0 +1,110 @@
+
+    def __init__(self, base_model):
+        """Initialize extended supernova model class."""
+        if not isinstance(base_model, SupernovaModel):
+            raise TypeError("ExtendedModel.__init__ requires a SupernovaModel object")
+
+        self.__dict__ = base_model.__dict__.copy()
+        for method_name in dir(base_model):
+            if callable(getattr(base_model, method_name)) and method_name[0] != '_':
+                if method_name == 'get_initial_spectra':
+                    self._get_initial_spectra = getattr(base_model, method_name)
+                else:
+                    self.method_name = getattr(base_model, method_name)
+        self.t_final = self.time[-1]
+        self.L_final = {Flavor.NU_E: self.luminosity[Flavor.NU_E][-1],
+                        Flavor.NU_X: self.luminosity[Flavor.NU_X][-1],
+                        Flavor.NU_E_BAR: self.luminosity[Flavor.NU_E_BAR][-1],
+                        Flavor.NU_X_BAR: self.luminosity[Flavor.NU_X_BAR][-1]}
+
+    def get_initial_spectra(self, *args):
+        """Get neutrino spectra/luminosity curves before oscillation"""
+        return self._get_initial_spectra(*args)
+
+    def get_extended_luminosity(self, t, k = -1., A = 1e51 * u.erg / u.s, tau_c = 36. * u.s, alpha = 2.66, flavor = Flavor.NU_E):
+        """Get neutrino luminosity from supernova cooling tail luminosity model.
+
+        Parameters
+        ----------
+        t : astropy.Quantity
+            Time to evaluate luminosity.
+        k : float
+            Power law factor (default: -1)
+        A : astropy.Quantity
+            Normalization factor (default: 1e51 erg/s)
+        tau_c : astropy.Quantity
+            Exponential decay characteristic timescale (default: 36 s)
+        alpha : float
+            Exponential decay factor (default: 2.66)
+
+        Returns
+        -------
+        astropy.Quantity
+            Luminosity calculated from cooling tail model.
+        """
+        if t.value < 0.5:
+            warn("Extended luminosity model not applicable to early times")
+        if A is None:
+            tf = self.t_final
+            Lf = self.L_final[flavor]
+            A = Lf/((tf.value**k) * np.exp(-1.*((tf/tau_c)**alpha)) * u.erg / u.s)
+        return A * (t.value**k) * np.exp(-1.*((t/tau_c)**alpha)) * u.erg / u.s
+
+    def extend(self, ts, k = -1., A = None, tau_c = 36. * u.s, alpha = 2.66):
+        """Extend supernova model to specific times.
+
+        Parameters
+        ----------
+        ts : astropy.Quantity
+            Times to add to supernova model.
+        k : float
+            Power law factor (default: -1)
+        A : astropy.Quantity
+            Normalization factor (default: None)
+        tau_c : astropy.Quantity
+            Exponential decay characteristic timescale (default: 36 s)
+        alpha : float
+            Exponential decay factor (default: 2.66)
+        """
+        for t in ts:
+            self.time = np.append(self.time, t)
```

Comment:
Should we enforce here that the times in `ts` are larger than `base_model.time[-1]`?

Or alternatively, if we’d like to include a larger time range in `ts` (e.g. to compare the “true” luminosity with the “extended” luminosity at early times), I think it would be better to keep the “true” and “extended” data points separate. If we just append the extended points to `self.time` (here) and `self.luminosity` (below), it’s more difficult to plot the true/extended luminosity separately and we get plotting artifacts, like in this example from the demo notebook:

![image](https://github.com/user-attachments/assets/139f9a59-183d-4a31-9e13-9d6541f6d70c)

Reply:
I’d … probably err on the side of not making the user keep track of such things; but I can also see how it might be lead to confusion if we try to be “too smart” and drop entries when the user might not expect that. 🤔

Maybe we should keep it as is for now, merge, and ask people to test a beta release? And then we can still change it based on feedback.